### DISCUSSION TITLE
FE 레벨4 SSR | OX퀴즈 해설지 만들기

### DISCUSSION THREAD

**Discussion by woowahan-cron** [woowacourse/retrospective#56] [General]
# 팀원과 OX퀴즈 해설지 만들기

각 항목별로 O 또는 X를 기입하고 왜 O 또는 X인지 예시를 들어 해설지를 만들어 봅시다

# 여러분 이 모든 문제는 본 미션 2단계를 완성했을 때를 기준으로 생각해 주세요! (사전 미션 고려 X)

- 아래 문제는,  <span style="color:red; font-weight:bold;">본 미션 저장소</span>의 요구사항 2단계를 모두 구현한 상태를 기준으로 합니다.

### 2. ~SSR에서는~본 미션에서는 초기 페이지 로드 시 웹 브라우저가 JS bundle 파일을 요청하지 않는다.
 
 (항목에 대한 양식은 자유롭게 작성합니다)

- 예시 답변) O
- 예시 해설) 

~SSR은~본 미션에서는 웹 브라우저에서 실행해야 할 모든 bundle 파일의 코드 내용을 서버에서 모두 실행하여 클라이언트에 결괏값만을 전달해 주기 때문에 실행 속도가 빠르며 웹 브라우저에서는 bundle 코드가 실행될 필요가 없다. 따라서 bundle 파일은 서버가 웹 브라우저로 서빙하지 않는다. 또한, 아래의 서버 코드는 클라이언트 대신 클라이언트에서 실행될 JS번들을 대신 실행해 주고 있다.

```js-light
app.get("/", async (req, res) => {
  const popularMovies = await loadMovies();
  const app = <ClientApp movies={popularMovies} />;

  const rendered = renderToString(app);
  const template = fs.readFileSync(path.join(clientBuildPath, "index.html"), "utf8");

  const initialData = /*html*/ `
    <script>
      window.__INITIAL_DATA__ = {
        movies: ${JSON.stringify(popularMovies)}
      }
    </script>
  `;

  res.send(template.replace('<div id="root"></div>', `<div id="root">${rendered}</div>${initialData}`));
});
```

### 3. 본 미션에서는 SSR을 사용하면 초기 페이지를 서버에서 첫 화면의 웹 문서를 완성시켜 온다. 따라서, 하이드레이션 **과정에서** 리액트에서는 첫 화면을 검증하지 않고 다음 렌더링을 준비한다.

-> 하이드레이션 **후** -> **과정**에서로 변경하여 문제를 풀어 주세요!

### 4. 본 미션에서는 SSR을 사용하면 초기 페이지를 서버에서 첫 화면의 웹 문서를 완성시켜 주기 때문에 CSR보다 빠른 렌더링 시간을 반드시 보장하고 있다.

**Comment by woowahan-cron** [comment]
# 팀 이름

- 구성원 (시모, 시모, 시모)

# 문제1

(해설)

# 문제2

(해설)

# 문제3

(해설)

# 문제4

(해설)

**Comment by chysis** [comment]
# 리뷰미
- 바다, 쑤쑤, 올리, 에프이

**미션 한정해서 풀었습니다**

# 문제1

### X
모달의 경우, 서버에서 json 형태의 데이터만 받아와서 리액트가 DOM을 그립니다. 따라서, 매번 새로운 HTML을 받아오는 것은 아닙니다.

# 문제2

### X
서버에서 렌더링된 HTML은 정적인 콘텐츠일 뿐, 클라이언트가 동적인 UI를 활성화하거나 사용자 상호작용을 처리하려면 자바스크립트가 필요합니다. 클라이언트 측에서 React가 동작하려면 자바스크립트 번들이 반드시 필요합니다. 이 번들이 로드되어야 클라이언트 측 React가 작동하고, 페이지가 상호작용 가능하게 변합니다.

# 문제3

### X
하이드레이션 과정에서 서버가 내려준 HTML과 React가 생성한 가상 DOM을 비교합니다. 이때 일치한다면 하이드레이션에 성공하고, 아니라면 실패합니다.

```text
react-dom.development.js:86 Warning: An error occurred during hydration. The server HTML was replaced with client content in <div>
react-dom.development.js:12507 Uncaught Error: Hydration failed because the initial UI does not match what was rendered on the server.
```
▲ 하이드레이션 실패 시 발생하는 오류 메시지

# 문제4

### X
- 서버 부하 등 네트워크 지연이 발생하는 경우
- API 호출이 오래 걸리는 경우

위의 두 경우에서는 SSR이 CSR보다 더 빠르다는 보장을 할 수 없습니다.

**Reply by woowahan-cron** [reply]
해설지를 작성하는 만큼 구체적인 과정이나 예시가 드러나면 좋겠어요 😄

**Comment by jinhokim98** [comment]
# 행동대장
+ 구성원 (토다리, 쿠키)

## 문제1
SSR을 사용하면 모든 사용자 상호작용에 대해 서버로부터 새로운 HTML을 받아와야 한다. (예를 들어, 영화 목록에서 사용자가 아이템을 클릭하면 다시 서버로부터 새로운 HTML을 요청한다)

답: X

SSR은 첫 페이지 로딩 시 서버에서 완성된 HTML을 받아오지만, 이후 사용자 상호작용은 클라이언트 측에서 처리됩니다. React는 하이드레이션 이후 클라이언트에서 작동하게 되며, 사용자가 상호작용할 때마다 서버로 새로운 HTML을 요청하지 않습니다. 따라서, 서버에서 매번 새로운 HTML을 받아오는 것이 아니라 클라이언트 측에서 JavaScript를 통해 인터랙션이 처리됩니다.


## 문제2
SSR에서는 초기 페이지 로드 시 웹 브라우저가 JS bundle 파일을 요청하지 않는다

답 : X

SSR을 사용하면 서버가 초기 HTML을 전달하지만, 이후 클라이언트는 JavaScript 번들 파일을 요청하여 하이드레이션 과정을 거칩니다. 이 번들 파일이 없다면 클라이언트 측에서의 상호작용이 불가능하기 때문에 번들 파일이 필요합니다.


## 문제3
SSR을 사용하면 초기 페이지를 서버에서 첫 화면의 웹 문서를 완성시켜 온다. 따라서, 하이드레이션 과정에서 리액트는 첫 화면을 검증하지 않고 다음 렌더링을 준비한다.

답 : O

하이드레이션 과정에서는 서버에서 받은 HTML과 React 상태가 일치하는지를 확인하지만, 하이드레이션 중 서버와 클라이언트의 상태가 대체로 일치할 경우 추가적인 검증 없이 바로 다음 렌더링을 준비합니다. 하이드레이션 과정 자체는 클라이언트에서 서버가 제공한 HTML을 활성화시키는 과정이며, 주로 개발 환경에서만 경고가 표시될 수 있지만 일반적으로 서버에서 내려온 HTML과 클라이언트의 상태가 일치하면 리액트는 다음 렌더링 단계로 넘어갑니다.


## 문제4
SSR을 사용하면 초기 페이지를 서버에서 첫 화면의 웹 문서를 완성시켜 주기 때문에 CSR보다 빠른 렌더링 시간을 반드시 보장하고 있다.

답 : X

SSR이 반드시 CSR보다 빠른 렌더링 시간을 보장하지 않습니다. SSR은 서버에서 HTML을 완성하여 보내기 때문에 첫 화면을 빨리 표시할 수 있지만, 서버에서 HTML을 생성하는 과정이 느리면 CSR보다 느릴 수 있습니다. CSR은 초기 로딩에 시간이 걸리지만, 이후의 상호작용이 더 빠르게 동작할 수 있습니다. 따라서, SSR이 항상 CSR보다 빠르다고 할 수는 없습니다.

**Comment by Parkhanyoung** [comment]
### 데벨업
구성원 (버건디, 프룬, 라이언)

**문제1**
X
미션과 같이 hydration을 수행한다고 하면 초기 로드 이후의 상호작용에 대해서는 CSR 방식으로 동작하기 때문에 새로운 페이지를 요청하지 않아도 된다.

**문제2**
세모
hydration을 수행한다면 번들을 요청할 것이고, 그렇지 않다면 요청하지 않을 것이다.

**문제3**
O
검증해서 서버가 내려준 결과물과 CSR 결과물이 다를 경우 에러메시지를 띄워준다. 단, 달라도 hydration은 진행된다.

**문제4**
X
초기 로드 시 보여줘야 하는 요소가 많을 경우, 서버에서 렌더링하는 데 오래 걸릴 수 있다. 반면 CSR은 초기 로드 시 보여줘야 하는 요소가 많아도 로딩 스피너나 페이지를 부분적으로 보여줄 수 있다.

**Comment by jaeml06** [comment]
# 팀 이름
- 구성원 (치코, 소파, 수야)
## 문제 1. SSR을 사용하면 모든 사용자 상호작용에 대해 서버로부터 새로운 HTML을 받아와야 한다.
- 해답: X
- 이유:
    - 치코버전: SSR을 사용하면 서버에서 만든 정적 HTML 파일을 반환한다. 하지만 모든 사용자 상호작용에 대해 새로운 HTML파일을 받아와야하지는 않는다. hover와 같은 css나 js로 상호작용 처리를 작성한 경우 새로운 HTML파일이 필요로하지 않는다.
    - 소파버전:CSS hover,js alert  등으로 인해 새로운 HTML을 안받아와도 된다
    - 수야버전: 모든 상호작용에 대해서 새로운 HTML을 받아오진 않는다. css나 js로 상호작용에 반응할 수 있다.

## 문제2. SSR에서는 초기 페이지 로드 시 웹 브라우저가 JS bundle 파일을 요청하지 않는다.
- 해답: x
- 이유: 하이드레이션 등을 위해서 파일을 요청할 수 있다.

## 문제3. SSR을 사용하면 초기 페이지를 서버에서 첫 화면의 웹 문서를 완성시켜 온다. 따라서, 하이드레이션 과정에서 리액트에서는 첫 화면을 검증하지 않고 다음 렌더링을 준비한다.
- 해답: X
- 이유: SSR을 사용하면, 서버에서 첫 화면의 HTML을 완성해서 브라우저에 전달한다. 이 HTML은 이미 서버에서 렌더링된 것이므로, 클라이언트에서는 이 HTML을 다시 그릴 필요가 없습니다. 대신 클라이언트에서는 Hydration 과정에서 해당 HTML에 React의 동적 기능을 결합한다.

## 문제4. SSR을 사용하면 초기 페이지를 서버에서 첫 화면의 웹 문서를 완성시켜 주기 때문에 CSR보다 빠른 렌더링 시간을 반드시 보장하고 있다.
- 해답: X
- 이유: 극단적으로 네트워크 환경이 느린 경우, 서버 스펙이 클라이언트 환경보다 떨어지는 경우에서 API 데이터를 가져오는데 많은 시간일 걸린다. 이럴경우 CSR에서 스켈레톤 UI를 보여주는 방식이 더 빠르게 사용자에게 보여질 수 있다.

**Reply by woowahan-cron** [reply]
팀 이름 -> 모우다로 변경해주세요 ㅋㅋ

**Comment by anttiey** [comment]
# 🍀 코딩해듀오
파란, 파슬리

## 1. SSR을 사용하면 모든 사용자 상호작용에 대해 서버로부터 새로운 HTML을 받아와야 한다. (예를 들어, 영화 목록에서 사용자가 아이템을 클릭하면 다시 서버로부터 새로운 HTML을 요청한다)

### X
SSR은 초기 페이지 로딩 시 서버에서 HTML을 생성하여 웹 브라우저로 전달하지만, 그 이후의 사용자 상호작용에 대해서는 꼭 서버로부터 새로운 HTML을 받아오지 않아도 됩니다. 클라이언트 측에서 JavaScript가 활성화된 경우 초기 페이지 로드 이후의 사용자 상호작용은 웹 브라우저에서 처리될 수 있습니다.

```js
const express = require('express');
const React = require('react');
const ReactDOMServer = require('react-dom/server');
const App = require('./App').default;

const server = express();

server.get('/', (req, res) => {
  const appString = ReactDOMServer.renderToString(<App />);
  
  res.send(`
    <!DOCTYPE html>
    <html>
      <head>
        <title>MovieList</title>
      </head>
      <body>
        <div id="root">${appString}</div>
        <script src="/static/client.js"></script>
      </body>
    </html>
  `);
});

server.use('/static', express.static('public'));

server.listen(3000, () => {
  console.log('Server is running on http://localhost:3000');
});
```
```js
import React, { useState } from 'react';

const App = () => {
  const [movies, setMovies] = useState(['Movie 1', 'Movie 2', 'Movie 3']);
  const [selectedMovie, setSelectedMovie] = useState(null);

  const handleClick = (movie) => {
    setSelectedMovie(movie);
  };

  return (
    <div>
      <h1>Movie List</h1>
      <ul>
        {movies.map((movie) => (
          <li key={movie} onClick={() => handleClick(movie)}>
            {movie}
          </li>
        ))}
      </ul>
      {selectedMovie && <div>Selected Movie: {selectedMovie}</div>}
    </div>
  );
};

export default App;
```

위 코드에서 서버는 초기 렌더링만 담당하고, 이후 사용자 상호작용은 리액트가 처리하고 있습니다.

## 2. SSR에서는 초기 페이지 로드 시 웹 브라우저가 JS bundle 파일을 요청하지 않는다.
### X
SSR을 사용하더라도 클라이언트 측의 상호 작용을 위해 JS 번들 파일이 필요합니다. 초기 HTML은 서버에서 렌더링되지만, 웹 브라우저에서 리액트와 같은 라이브러리가 하이드레이션 과정을 통해 페이지가 상호 작용하게 합니다. 이 과정에서 JS 번들이 필요합니다.

```js
import express from 'express';
import React from 'react';
import ReactDOMServer from 'react-dom/server';
import App from './src/App';

const app = express();

app.get('/', (req, res) => {
  const appString = ReactDOMServer.renderToString(<App />);
  
  res.send(`
    <!DOCTYPE html>
    <html>
      <head>
        <title>SSR Count</title>
      </head>
      <body>
        <div id="root">${appString}</div>
        <script src="/static/bundle.js"></script>
      </body>
    </html>
  `);
});

app.use('/static', express.static('dist'));

app.listen(3000, () => {
  console.log('Server is running on http://localhost:3000');
});
```
```js
import React, { useState } from 'react';

const App = () => {
  const [count, setCount] = useState(0);

  return (
    <div>
      <h1>Hello, SSR!</h1>
      <p>Count: {count}</p>
      <button onClick={() => setCount(count + 1)}>Increment</button>
    </div>
  );
};

export default App;
```

위 예시에서 서버의 HTML을 통해 초기 렌더링을 진행하지만, 버튼 클릭 시 상호 작용은 JS 번들 파일을 전달 받은 웹 브라우저의 하이드레이션 과정을 통해 추가됩니다.

## 3. SSR을 사용하면 초기 페이지를 서버에서 첫 화면의 웹 문서를 완성시켜 온다. 따라서, 하이드레이션 과정에서 리액트에서는 첫 화면을 검증하지 않고 다음 렌더링을 준비한다.
### X

```bash
react-dom.development.js:86 Warning: An error occurred during hydration. The server HTML was replaced with client content in <div>
react-dom.development.js:12507 Uncaught Error: Hydration failed because the initial UI does not match what was rendered on the server.
```

위 오류 예시에서도 나타나듯이, 리액트는 하이드레이션 과정에서 `서버에서 전달받은 HTML` 과 `리액트 컴포넌트` 가 서로 같은지 검증하고, 다르면 경고 혹은 오류를 반환합니다. 둘의 구조가 서로 다를 경우, 개발자가 의도한 부분에 의도한 상호 작용이 일어나지 않을 가능성이 높기 때문에 리액트에서는 하이드레이션 과정에서 반드시 화면을 검증합니다.

## 4. SSR을 사용하면 초기 페이지를 서버에서 첫 화면의 웹 문서를 완성시켜 주기 때문에 CSR보다 빠른 렌더링 시간을 반드시 보장하고 있다.
### X

SSR이 초기 페이지 로드에 유리한 방식은 맞지만, 항상 CSR보다 빠른 렌더링 시간을 보장하는 것은 아닙니다.

1. 서버의 처리 시간이 데이터베이스 요청이나 복잡한 연산 등으로 길어지는 경우, 이로 인해 렌더링 속도가 느려질 수 있습니다.
2. 서버에서 생성된 HTML을 클라이언트로 전송하는 동안 네트워크 지연이 발생할 수 있습니다.

**Reply by woowahan-cron** [reply]
3번 문제 정정: **하이드레이션 과정에서**로 수정하여 진행해주세요!

**Comment by Largopie** [comment]
# 모모
구성원 (해낙빙)

## **1. 미션에서 ~~SSR을 사용하면(제외)~~ 모든 사용자 상호작용에 대해 서버로부터 새로운 HTML을 받아와야 한다. (예를 들어, 영화 목록에서 사용자가 아이템을 클릭하면 다시 서버로부터 새로운 HTML을 요청한다)**

정답: X

해설: 미션에서는 Hydration 이후 상호작용이 발생하면 새로운 HTML을 받아오지 않고 클라이언트 측에서 JavaScript로 상호작용할 수 있다.

## ~~2. SSR에서는~~본 미션에서는 초기 페이지 로드 시 웹 브라우저가 JS bundle 파일을 요청하지 않는다.

정답: X

해설: Hydration을 위해 JS bundle 파일을 요청한다.

## **3. 본 미션에서는 SSR을 사용하면 초기 페이지를 서버에서 첫 화면의 웹 문서를 완성시켜 온다. 따라서, 하이드레이션 과정에서 리액트에서는 첫 화면을 검증하지 않고 다음 렌더링을 준비한다.**

정답: X

해설: (앞 문장 맞음). **하이드레이션 과정에서 리액트에서는 첫 화면을 ~~검증하지 않고~~ 검증하고 일치하면 다음 렌더링을 준비한다. 일치하지 않으면 hydration failed 에러 발생**

![image](https://github.com/user-attachments/assets/cab5f086-e89d-4bb5-81a5-5252a5e377ab)


## 4. 본 미션에서는 SSR을 사용하면 초기 페이지를 서버에서 첫 화면의 웹 문서를 완성시켜 주기 때문에 CSR보다 빠른 렌더링 시간을 반드시 보장하고 있다.

정답: X

해설: 반드시는 아니다. 서버에 부하가 생겨서 요청 시간이 오래걸린다면 CSR이 더 빠를 수 있다. 왜냐하면 CSR은 데이터 fetching은 페이지 렌더링 이후에 하기 때문이다.

**Comment by chlwlstlf** [comment]
# 코레아
다르, 텐텐, 초코

# 문제1
X
첫 페이지만 HTML을 받아오고 그 후에 사용자의 상호작용으로 바뀌는 데이터는 ajax로 처리한다.

# 문제2
X
초기 렌더링 이후에 CSR이기 때문에 JS bundle 파일을 요청한다.

# 문제3
X
서버에서 받은 HTML 파일을 파싱한 DOM 트리와 React가 그리는 DOM 트리가 일치하는 지 확인한다. 이 과정에서 두 개가 일치하지 않으면 에러가 뜬다.

# 문제4
X
외부 서버에 의존하기 때문에 반드시는 아니다. 
TMDB 서버 응답을 10초로 가정한다면 SSR보다 CSR에서의 초기 렌더링이 훨씬 빠르다. 그 이유는 SSR은 데이터를 받아온 후 html을 완성시켜서 클라이언트로 보내기 때문에 10초 이상이 걸릴 것이고, CSR에서는 정적 파일을 불러와서 TMDB 서버 페칭 전까지의 UI를 볼 수 있기 때문에 CSR에서 훨씬 빠르게 초기 페이지를 볼 수 있다.

**Comment by Hain-tain** [comment]
# 코드잽
구성원 (헤인, 마위, 월하)

## 문제1
- Q:  본 미션에서 모든 사용자 상호작용에 대해 서버로부터 새로운 HTML을 받아와야 한다. (예를 들어, 영화 목록에서 사용자가 아이템을 클릭하면 다시 서버로부터 새로운 HTML을 요청한다)
- A: X. 초기 렌더링시에만 서버로부터 HTML을 받아오고, 그 이후 상호작용에 대해서는 csr 로 작동한다.

## 문제2
- Q: 본 미션에서 SSR에서는 초기 페이지 로드 시 웹 브라우저가 JS bundle 파일을 요청하지 않는다.
- A: X, HTML 파일 내부에 스크립트 태그로 bundle.js 를 가져오기때문에 요청한다.

## 문제3
- Q:  본 미션에서는 SSR을 사용하면 초기 페이지를 서버에서 첫 화면의 웹 문서를 완성시켜 온다. 따라서, 하이드레이션 과정에서 리액트에서는 첫 화면을 검증하지 않고 다음 렌더링을 준비한다.
- A: X, 하이드레이션 랜더링된 화면과 가상 돔의 데이터와 검증하는 단계를 포함하기 때문에 첫 화면을 검증한다.

## 문제4
- Q: 본 미션에서는 SSR을 사용하면 초기 페이지를 서버에서 첫 화면의 웹 문서를 완성시켜 주기 때문에 CSR보다 빠른 렌더링 시간을 반드시 보장하고 있다.
- A: ▵, 보통의 경우 SSR이 CSR보다 빠른 렌더링 시간을 보장한다고 생각한다.

**Comment by healim01** [comment]
# 방끗
헤일리, 리안, 제이드

# 문제1
### 1. SSR을 사용하면 모든 사용자 상호작용에 대해 서버로부터 새로운 HTML을 받아와야 한다. (예를 들어, 영화 목록에서 사용자가 아이템을 클릭하면 다시 서버로부터 새로운 HTML을 요청한다)

답변 : X 
2단계 미션 특성상 초기 랜더시에만 HTML 파일을 받아오고, 이후에는 CSR 형식으로 데이터를 요청한다. 



# 문제2
### 2. SSR에서는 초기 페이지 로드 시 웹 브라우저가 JS bundle 파일을 요청하지 않는다.

답변 : X 
받아온 HTML 파일 내부에 스크립트로 bundle.js 를 불러오기 때문에 요청하게 된다. 


# 문제3
### 3. SSR을 사용하면 초기 페이지를 서버에서 첫 화면의 웹 문서를 완성시켜 온다. 따라서, 하이드레이션 과정에서 리액트에서는 첫 화면을 검증하지 않고 다음 렌더링을 준비한다.

답변 : X 
하이드레이션 자체가 랜더링된 화면과 가상 돔의 데이터와 검증하는 단계를 포함하기 때문에 첫 화면을 검증한다.

# 문제4
### 4. 본 미션에서 SSR을 사용하면 초기 페이지를 서버에서 첫 화면의 웹 문서를 완성시켜 주기 때문에 CSR보다 빠른 렌더링 시간을 반드시 보장하고 있다.

답변 : ▵
SSR 은 초기 페이지를 서버에서 완성 시키기 때문에 CSR 보다는 빠른 랜더링 시간을 보장한다. 단 반드시 일지는 미지수

**Comment by llqqssttyy** [comment]
## 크루루
구성원 (렛서, 아르, 러기)

### 1. SSR을 사용하면 모든 사용자 상호작용에 대해 서버로부터 새로운 HTML을 받아와야 한다. 
- X
(예를 들어, 영화 목록에서 사용자가 아이템을 클릭하면 다시 서버로부터 새로운 HTML을 요청한다) 상황마다 다름

### 2. SSR에서는 초기 페이지 로드 시 웹 브라우저가 JS bundle 파일을 요청하지 않는다. 
- X
상황마다 다르다. script태그를 삽입할 때 src속성을 넣으면 요청하고, 아니면 요청 안합니다.

### 3. SSR을 사용하면 초기 페이지를 서버에서 첫 화면의 웹 문서를 완성시켜 온다. 따라서, 하이드레이션 과정에서 리액트에서는 첫 화면을 검증하지 않고 다음 렌더링을 준비한다.  
- X
검증한다. Hyration Error 발생한거 보면 그렇더라. 리액트의 이벤트 구조, 렌더링 처리 방식에 따른 예상치 못한 동작을 야기할 수 있다.

### 4. SSR을 사용하면 초기 페이지를 서버에서 첫 화면의 웹 문서를 완성시켜 주기 때문에 CSR보다 빠른 렌더링 시간을 반드시 보장하고 있다. 
- X
반드시는 아니지만, 대다수은 서버에서 csr보단 빠르게 코드를 처리할 수 있다.

**Comment by jinyoung234** [comment]
# 투룻

- 리버, 지니, 시모

# **1. 모든 사용자 상호작용에 대해 서버로부터 새로운 HTML을 받아와야 한다. (예를 들어, 영화 목록에서 사용자가 아이템을 클릭하면 다시 서버로부터 새로운 HTML을 요청한다)**

정답 - x

이유 - ‘모든’이라는 단어가 잘못되었다. 왜냐하면 사용자 상호작용에 의해 새로운 페이지로 라우팅 된 경우(ex - click 핸들러에 navigate(’other-page’)를 등록한 경우) 만약 SSR 처리(특정 path에 대한 서버 처리)를 했다면 새로운 html을 받게 된다.

하지만 그 외의 동작의 경우 초기 정적 렌더링 후 번들 파일(js)이 실행 되었기 때문에 그 이후 사용자 액션에 대해서는 csr에 의해 제어되어 새로운 html이 요청되지 않는다.

# **2. 본 미션에서는 초기 페이지 로드 시 웹 브라우저가 JS bundle 파일을 요청하지 않는다.**

정답 - x

이유 - 빌드 시 webpack에 의해 번들 파일을 index.html에 포함시키기 때문에 SSR로 초기 페이지를 서빙한다고 하더라도 브라우저 렌더링 과정에서 번들파일의 script를 읽기 때문에 bundle 파일을 요청한다.

# 3. **SSR을 사용하면 초기 페이지를 서버에서 첫 화면의 웹 문서를 완성시켜 온다. 따라서, 하이드레이션 과정에서 리액트에서는 첫 화면을 검증하지 않고 다음 렌더링을 준비한다.**

정답 - x

```js
react-dom.development.js:86 Warning: An error occurred during hydration. The server HTML was replaced with client content in <div>
react-dom.development.js:12507 Uncaught Error: Hydration failed because the initial UI does not match what was rendered on the server.
```

이유 - hydrateRoot에 의해 하이드레이션 과정에서 서버에서 내려준 html과 리액트 트리를 검증하는 과정을 거쳐서 같은 구조를 가지지 않는 경우 에러를 던진다. 따라서 하이드레이션 과정에서 검증하지 않는다는 표현은 틀린 표현이다.

# **4. SSR을 사용하면 초기 페이지를 서버에서 첫 화면의 웹 문서를 완성시켜 주기 때문에 CSR보다 빠른 렌더링 시간을 반드시 보장하고 있다.**

정답 - x

```js
function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

router.get('/', async (_, res) => {
  currentMovieType = 'nowPlaying';

  await sleep(10000);

  const movies = await fetchMovieList();

  res.send(renderHTML(movies));
});
```


이유 - 서버가 부하가 발생하는 경우에는 발생하는 시간 만큼 초기 페이지가 화면에 노출되지 않기 때문에 CSR보다 느려질 수 있다. 하지만 대부분은 SSR이 CSR이 빠른 편이다.

**Comment by novice0840** [comment]
# 땅콩
# 구성원 (포메, 썬데이, 마루)

# 문제1
상호작용은 CSR과 SSR 양쪽을 모두 사용하므로
모든 경우를 SSR로 렌더링하지는 않는다.

서버에서 만들어진 HTML에 리액트 컴포넌트 로직을 결합하는 형태이기 때문에 상호작용에는 리액트의 CSR 방식이 사용된다. 따라서 아이템에 대한 JSON 데이터를 응답받아 기존의 HTML을 갈아끼운다.

SSR + CSR이 결합된 형태에서는 상호작용이 발생했을 때 전체 페이지를 다시 로드하지 않고 필요한 부분만 업데이트한다.


# 문제2
React로 만든 앱과 상호작용하기 위해서는 bundle.js 파일이 필요하다. 실제로 네트워크 탭을 보면 bundle 파일을 불러오는 것을 확인할 수 있다.

서버에서 만든 HTML은 정적인 파일이다. bundle 파일을 불러와 hydration이 적용돼야 상호작용이 가능하게 된다. 따라서 브라우저에서 HTML을 불러올 때 bundle 파일을 요청하여 hydration을 진행한다.

미션에서 실제 네트워크 탭에서 bundle.js 파일을 불러오는 것을 확인할 수 있다.bundle 불러와  dydration이 되면 사용자와 상호작용을 할 수 있다.

# 문제3
SSR로 만든 HTML과 React로 만든 App의 일치 여부를 확인하고 HTML에 상호작용을 추가하기 때문에 화면 검증이 필요하다.

서버에서 만든 HTML과 리액트 트리가 일치해야 hydration이 정상적으로 된다. 검증에 실패한다면 화면을 불필요하게 다시 그리거나, DOM 구조가 다르다면 정상적으로 hydration이 되지 않는다.

첫 화면은 검증하지 않는 것이 아니고 화면을 그릴 때 hydration의 과정에서 서버와 클라이언트의 DOM 트리가 일치하는지 확인한다.

# 문제4
대부분의 경우에는 SSR이 빠르지만 반드시 보장되지는 않는다

SSR 서버의 부하가 크다면 브라우저에서 요청하는 것보다 느릴 수 있다. 따라서 반드시 빠른 렌더링 시간을 보장하지는 않는다.

렌더링 속도에는 서버의 응답 속도, 네트워크 속도 등 여러 요인이 포함되기 때문에 초기 로딩 시간이 SSR이 빠르다고 해서 더  빠른 렌더링 시장을 보장하지는 않는다.
